theano/gpuarray/linalg.py

```diff
@@ -503,3 +491,116 @@ def gpu_matrix_inverse(a):
 
     """
     return GpuMagmaMatrixInverse()(a)
+
+
+class GpuMagmaCholesky(GpuMagmaBase, CGpuKernelBase):
+    """Computes the cholesky decomposition of a matrix :math:`A` using magma
+    library.
+
+    """
+    __props__ = ('lower', 'inplace')
+
+    def __init__(self, lower=True, inplace=False):
+        self.lower = lower
+        COp.__init__(self, ['magma_cholesky.c'], 'APPLY_SPECIFIC(magma_cholesky)')
+        self.inplace = inplace
+        if self.inplace:
+            self.destroy_map = {0: [0]}
+
+    def clone_inplace(self):
+        return self.__class__(lower=self.lower, inplace=True)
+
+    def make_node(self, A):
+        ctx_name = infer_context_name(A)
+        A = as_gpuarray_variable(A, ctx_name)
+        A = gpu_contiguous(A)
+        if A.ndim != 2:
+            raise LinAlgError("Matrix rank error")
+        return theano.Apply(self, [A], [A.type()])
+
+    def get_op_params(self):
+        params = []
+        if self.lower:
+            params.append(('LOWER', '1'))
+        if self.inplace:
+            params.append(('INPLACE', '1'))
+        return params
+
+    def infer_shape(self, node, shapes):
+        return [shapes[0]]
+
+
+class GpuMagmaQR(GpuMagmaBase, CGpuKernelBase):
+    """Computes the qr decomposition of a matrix :math:`A` using magma
+    library.
+
+    Parameters
+    ----------
+    complete : If `False`, returns only r.
+
+    """
+    __props__ = ('complete', )
+
+    def __init__(self, complete=True):
+        self.complete = complete
+        COp.__init__(self, ['magma_qr.c'], 'APPLY_SPECIFIC(magma_qr)')
+
+    def make_node(self, A):
+        ctx_name = infer_context_name(A)
+        A = as_gpuarray_variable(A, ctx_name)
+        A = gpu_contiguous(A)
+        if A.ndim != 2:
+            raise LinAlgError("Matrix rank error")
+        if self.complete:
+            return theano.Apply(self, [A], [A.type(), A.type()])
+        else:
+            return theano.Apply(self, [A], [A.type()])
+
+    def get_op_params(self):
+        params = []
+        if self.complete:
+            params.append(('COMPLETE', '1'))
+        return params
+
+
+class GpuMagmaEigh(GpuMagmaBase):
+    """Computes the eigen decomposition of a symmetric matrix :math:`A` using magma
+    library.
+
+    Parameters
+    ----------
+    UPLO : Specifies whether the calculation is done with the lower triangular
+           part of matrix (`L`, default) or the upper triangular part (`U`).
+    compute_v : If `True`, computes eigenvalues and eigenvectors (`True`,
+                default). If `False`, computes only eigenvalues of matrix.
+    """
+    __props__ = ('lower', )
+
+    def __init__(self, UPLO='L', compute_v=True):
+        assert UPLO in ['L', 'U']
+        self.lower = UPLO == 'L'
+        self.compute_v = compute_v
+        COp.__init__(self, ['magma_eigh.c'], 'APPLY_SPECIFIC(magma_eigh)')
+
+    def make_node(self, A):
+        ctx_name = infer_context_name(A)
+        A = as_gpuarray_variable(A, ctx_name)
```

Comment:
As you require A to be C contiguous in C code (magma_eigh.c), you should add `A = gpu_contiguous(A)` after this line.